apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ClientCacheExtensions.kt

```diff
@@ -95,14 +99,83 @@ fun ApolloClient.Builder.store(store: ApolloStore, writeToCacheAsynchronously: B
 }
 
+enum class WatchErrorHandling {
+  THROW_CACHE_ERRORS,
+  THROW_NETWORK_ERRORS,
+  THROW_CACHE_AND_NETWORK_ERRORS,
+  IGNORE_ERRORS
+}
+
 /***
  * Gets the result from the network, then observes the cache for any changes.
  * Overriding the [FetchPolicy] will change how the result is first queried.
+ * Exception are ignored by default, this can be changed by setting [fetchErrorHandling] for the first fetch and [refetchErrorHandling]
+ * for subsequent fetches.
+ */
+@JvmOverloads
+fun <D : Query.Data> ApolloCall<D>.watch(
+    fetchErrorHandling: WatchErrorHandling = WatchErrorHandling.IGNORE_ERRORS,
+    refetchErrorHandling: WatchErrorHandling = WatchErrorHandling.IGNORE_ERRORS,
+): Flow<ApolloResponse<D>> {
+  var data: D? = null
+  return toFlow()
+      .catch {
+        maybeThrow(it, fetchErrorHandling)
+      }.onEach {
+        data = it.data
+      }.onCompletion {
+        emitAll(
+            watch(data)
+                .catch {
+                  maybeThrow(it, refetchErrorHandling)
+                }
+        )
+      }
+}
+
+/**
+ * Observes the cache for the given data. Unlike [watch], no initial request is executed on the network.
+ * The refetch policy set by [refetchPolicy] will be used.
+ * Cache and network exceptions are propagated.
  */
-fun <D : Query.Data> ApolloCall<D>.watch(): Flow<ApolloResponse<D>> {
-  return copy().addExecutionContext(WatchContext(true)).toFlow()
+fun <D : Query.Data> ApolloCall<D>.watch(data: D?): Flow<ApolloResponse<D>> {
+  return copy().addExecutionContext(WatchContext(data)).toFlow()
 }
 
+private fun maybeThrow(throwable: Throwable, errorHandling: WatchErrorHandling) {
+  // Only potentially swallow Apollo exceptions, the rest must be surfaced
+  if (throwable !is ApolloException) {
+    throw throwable
+  }
+  if (errorHandling == WatchErrorHandling.IGNORE_ERRORS) return
+  val throwCacheErrors = errorHandling == WatchErrorHandling.THROW_CACHE_AND_NETWORK_ERRORS || errorHandling == WatchErrorHandling.THROW_CACHE_ERRORS
+  val throwNetworkErrors = errorHandling == WatchErrorHandling.THROW_CACHE_AND_NETWORK_ERRORS || errorHandling == WatchErrorHandling.THROW_NETWORK_ERRORS
+  when (throwable) {
+    is ApolloCompositeException -> {
+      if (errorHandling == WatchErrorHandling.THROW_CACHE_AND_NETWORK_ERRORS) {
+        throw throwable
+      }
+      val cacheMissException = throwable.first as? CacheMissException ?: throwable.second as? CacheMissException
+      // If it's *not* a CacheMissException we consider it a network error (could be ApolloNetworkException, ApolloHttpException, ApolloParseException...)
+      val networkException = throwable.first.takeIf { it !is CacheMissException } ?: throwable.second.takeIf { it !is CacheMissException }
+      if (cacheMissException != null && throwCacheErrors) {
+        throw cacheMissException
+      }
+      if (networkException != null && throwNetworkErrors) {
+        throw networkException
+      }
```

Comment:
I'm wondering if the `ApolloCompositeException` case is really required here. Could we instead consider `ApolloCompositeException` a network error? At the end of the day,  when `ApolloCompositeException` is thrown, it always means there was a network error (and maybe a cache error too). It keeps things simple. If anyone wants more fined grained error handling, they can always use the low-level API?

Reply:
Let's keep it as parameters then? To highlight the fact that it's more of a special case?